pkg/archive/archive.go

```diff
@@ -530,7 +550,8 @@ func createTarFile(path, extractDir string, hdr *tar.Header, reader io.Reader, L
 		}
 
 	case tar.TypeLink:
-		targetPath := filepath.Join(extractDir, hdr.Linkname)
+		lpath := adjustHardLink(hdr.Linkname, hdr.Name)
+		targetPath := filepath.Join(extractDir, lpath)
```

Comment:
@cyphar, For docker cp, I think the bug  is in  the generation or Untar, but in some other case，the bug is only in Untar. For  example, we save the  tar data(with hard link) to test.tar.gz on the docker daemon when we execute docker cp, or just use tar command to save to test.tar.gz,to uncompress test.tar.gz, the following code doesn't work,
```
package main

import (
	"os"
	"fmt"
	
	"github.com/docker/docker/pkg/archive"
)

func main() {
	in, _ := os.OpenFile(os.Args[1], os.O_RDWR, 0666)
	srcInfo := archive.CopyInfo{
		Path:       "/tmp",
		Exists:     true,
		IsDir:      true,
	}
	if err := archive.CopyTo(in, srcInfo, os.Args[2]); err != nil{
		fmt.Printf("%v\n", err)
	}
}
```
```
root@ubuntu:~/root/tar# go build cp.go
root@ubuntu:~/root/tar# cd tmp/
root@ubuntu:~/root/tar# echo "aaaa">a
root@ubuntu:~/root/tar# ln a b
root@ubuntu:~/root/tar# cd ../
root@ubuntu:~/root/tar# tar xvf test.tar.gz tmp
tmp/
tmp/b
tmp/a

./cp test.tar.gz tmp__
```
after this pr, the code above will work, and I think the change is forward compatible.